### DISCUSSION TITLE
Named placeholders for SQL

### DISCUSSION THREAD

**Discussion by johnfoldager** [JSQLParser/JSqlParser#1467] [Q&A]
I have a string representing a SQL statement with placeholders (`:name-of-placeholder`) like:

`SELECT * FROM table WHERE id = :id AND name = :name AND description LIKE "%:something%"`

Is it possible to process it so that I can get each placeholder (`:id` and `:name`, but not `:something`) and replace each with a value (numeric or string), such that:

- numerics will be inserted in-place without quotes,
- strings will be inserted in-place with quotes around but in such a way so that if string is fx `' OR TRUE'` then ti should escape it correct in order to prevent SQL injection.

**Comment by manticore-projects** [comment]
Greetings!

Yes, this is certainly possible, however I wonder if you are not looking for a more sophisticated approach:
1) identify all named parameters
2) replace with `?` placeholder for un-named parameters
3) create a map between every named parameter and its position of un-named parameters
4) provide an utility function, which fills the values of the named parameters for each un-named parameter using `setObject( position, value)`

Rationale: this approach allows for using Prepared Statements and Batch Updates yielding in much better performance instead of parsing single statements with hard coded parameters under heavy performance penalty.

In a nutshell, you would call a methods:

1) `ResultSet rs = getResultSet( String sqlStr, Map<String, Object> parameters)` or
2) `long n = executeUpdate(String sqlStr, Map<String, Object> parameters)`, which will depend on a Map/Pool of Prepared Statements identified by the SQL ID

This is not necessary in Scope of JSQLParser, but rather belongs into a JDBC extension.

**Reply by johnfoldager** [reply]
The solution to call a method with the SQL string and a map of parameters is what I need, however, I haven't been able to figure out how to do this the correct way. I've already tried the following for named queries:
- NamedPreparedStatement from https://www.infoworld.com/article/2077706/named-parameters-for-preparedstatement.html
- NamedParameterStatement from https://www.codemeright.com/blog/post/named-parameterized-query-java

but both fail on SQL injection problems. This is why I searched and found JSqlParser which might be able to actually find these `:parameters` and replace with a value and at the same time know how to escape quotes etc correctly in order to prevent SQL injection.

To me it looks like something like https://github.com/JSQLParser/JSqlParser/wiki/Examples-of-SQL-building--part-2 could potentially be used for this, however, I haven't worked with JSqlParser before, so wanted to know if this is possible at all.

**Reply by johnfoldager** [reply]
By the way.... forgot to say that the SQL comes from a configuration file and is made by another system based on different stuff, so I have to somehow parse SQL correctly, identify these `:placeholder`s and replace based on some values that I have.

**Reply by manticore-projects** [reply]
Yes, clear. I will send you a few lines of code tomorrow.

**Comment by manticore-projects** [comment]
Greetings.

A first commit: https://github.com/manticore-projects/MJdbcUtils
Have a look at the Unit Tests, they illustrate how to use it.

I will add more stuff over the next couple of days.

**Reply by johnfoldager** [reply]
Looks promising and in a secure environment it would work, but if you try the following test:

```
    @Test
    void rewriteStatementWithNamedParameters() throws Exception {
        Date dateParameterValue = new Date();

        HashMap<String, Object> parameters = new HashMap<>();
        parameters.put("param1", "' or 'A' <> 'B");

        String sqlStr = "select * from table_a where field_a = :param1";
        String rewrittenSqlStr = MJdbcTools.rewriteStatementWithNamedParameters(sqlStr, parameters);
        String correctSqlStr = "SELECT * FROM table_a WHERE field_a = ''' OR ''A'' <> ''B'";

        System.out.println("rewrittenSqlStr:  " + rewrittenSqlStr);
        System.out.println("correctEndSqlStr: " + correctSqlStr);

        Assertions.assertEquals(correctSqlStr, rewrittenSqlStr);
    }
```

then you can see the outpus is:

```
rewrittenSqlStr:  SELECT * FROM table_a WHERE field_a = '' or 'A' <> 'B'
correctEndSqlStr: SELECT * FROM table_a WHERE field_a = ''' OR ''A'' <> ''B'

expected: <SELECT * FROM table_a WHERE field_a = ''' OR ''A'' <> ''B'> but was: <SELECT * FROM table_a WHERE field_a = '' or 'A' <> 'B'>
Expected :SELECT * FROM table_a WHERE field_a = ''' OR ''A'' <> ''B'
Actual   :SELECT * FROM table_a WHERE field_a = '' or 'A' <> 'B'
```

The problem is that param1 is 'just a string' so it should be escaped correctly so we eliminate SQL injections.

**Reply by manticore-projects** [reply]
Thanks for testing it John. I will take your input and smoother the edges.
But please pay attention:
```java
parameters.put("param1", "' or 'A' <> 'B");
```

This is not an example of a Name Parameter! Instead, it is SQL rewriting, which is not in scope and a completely different use case. A Parameter really can take only one value, but must not contain or alter SQL query logic.

**Reply by johnfoldager** [reply]
Maybe the problem can be solved by just replacing single apostrophes with double ones before setting the string.

**Reply by johnfoldager** [reply]
Just tested this:
```
parameters.put("param1", "' or 'A' <> 'B".replaceAll("'", "''"));
```
and it works. However, I still think that a SQL API should handle this internally so that users do not forget SQL injections.

**Reply by manticore-projects** [reply]
> Maybe the problem can be solved by just replacing single apostrophes with double ones before setting the string.

It is really not about the quotes, but about the use case:

1) Do you want to execute `PreparedStatements` with (different sets of) parameters 
2) OR do you want to build various different SQL statements based on a template

I focus on 1) and would advise to look for a real Template Engine (like ThymeLeaf), if you are more interested in 2).
The provided example is about 2) Templating.

**Reply by johnfoldager** [reply]
What I am primarily interested in is:

1. Parse a string (which includes parameter names like `:param1`) and validate if it correct SQL syntax
2. Parse a string (which includes parameter names like `:param1`) and get a list of all parameter names and preferable where they are located in the string

and if both above succeeds then:

- Replace named parameters with different values and get a correct SQL syntax that a) escapes strings correctly, and b) makes sure to not replace fx `:param1` if this is inside a string like `'...:param1...'`.

**Reply by manticore-projects** [reply]
Yes, 1) or 2) is ok and covered -- but altering the SQL AST is a different thing and out of scope of SQL Named Parameters.
Your value does not provide a simple Parameter Value, but alters the meaning of the SQL itself.

Anyway, it is not strictly a "black or white" problem. You got your example working, I just point out that you are tampering with the AST and this will not work with PreparedStatements and BatchUpdates.

**Reply by johnfoldager** [reply]
I don't agree ;-)

I'm setting a parameter to a string which is put into the SQL statement without being properly escaped. As I wrote I can escape the parameter before putting it into the SQL. The problem is that the parameter value itself can change the SQL resulting in a different AST. However, if each string parameters was escaped correctly then it would result in the same AST no matter if that parameter did or did not try SQL injection.

**Reply by johnfoldager** [reply]
I might be wrong, and I'm going to look deeper into it tomorrow. For now I can replace string apostrophes with double ones to prevent SQL injection.

**Reply by manticore-projects** [reply]
Can you please give me the final SQL, properly formatted and with syntax highlighting? Just as a plain SQL without any Java code? I guess I got confused and saw a SQL Injection, which you actually want to avoid.

**Reply by manticore-projects** [reply]
I think, I completely misunderstood your request, sorry for that.
I have added an Oracle DB compliant escaping for `'` and `&`. Give it a try please.

**Reply by johnfoldager** [reply]
Now your code works as expected which indicates to me that JSqlParser with some add-on code can be used for my little project. However, since your code is GPLv3 licenses I don't think I can use it in my project.

**Comment by manticore-projects** [comment]
I have pushed support for `PreparedStatements`, which will effectively avoid SQL injection:
```sql
CREATE TABLE test ( 
a DECIMAL(3) PRIMARY KEY
, b VARCHAR(128) NOT NULL
, c DATE NOT NULL
, d TIMESTAMP NOT NULL
, e DECIMAL(23,5) NOT NULL 
) 
```

```java
String ddlStr = "INSERT INTO test VALUES ( :a, :b, :c, :d, :e )";
String qryStr = "SELECT Count(*) FROM test WHERE a = :a or b = :b";

Map<String, Object> parameters = toMap("a", 1, "b", "Test String", "c", new Date(), "d", new Date(), "e", "0.12345");

// Insert data
MPreparedStatement st = new MPreparedStatement(conn, ddlStr);
st.execute(parameters);

// Query data
MPreparedStatement st = new MPreparedStatement(conn, qryStr);
ResultSet rs = st.executeQuery(parameters);
```

**Reply by manticore-projects** [reply]
Added support for Batch execution:
```java
                MPreparedStatement st = new MPreparedStatement(conn, ddlStr, batchSize);

                for (int i=0; i < maxRecords; i++) {
                    parameters.put("a", i);
                    parameters.put("b", "Test String " + i);
                    
                    // effectively executed only when BATCH_SIZE is reached
                    int[] results = st.addAndExecuteBatch(parameters);
                }
                // effectively executed only when there are uncommitted records left
                int[] results = st.executeBatch();
```
